### PR TITLE
495 fix error: incompatible with supertype

### LINKED ISSUE #495: Mypy warning: Signature of "__call__" incompatible with supertype "Transform"
**Describe the bug**
Mypy reported warnings related to inconsistent function prototypes.

**To Reproduce**
1. Remove `# type: ignore` suppression comments from 
```
monai/transforms/utility/array.py:121: error: Signature of "__call__" incompatible with supertype "Transform"
monai/transforms/utility/array.py:164: error: Signature of "__call__" incompatible with supertype "Transform"
monai/transforms/utility/array.py:212: error: Signature of "__call__" incompatible with supertype "Transform"
monai/transforms/spatial/array.py:82: error: Signature of "__call__" incompatible with supertype "Transform"
monai/transforms/spatial/array.py:165: error: Signature of "__call__" incompatible with supertype "Transform"
monai/transforms/spatial/array.py:264: error: Signature of "__call__" incompatible with supertype "Transform"
monai/transforms/spatial/array.py:332: error: Signature of "__call__" incompatible with supertype "Transform"
monai/transforms/spatial/array.py:474: error: Signature of "__call__" incompatible with supertype "Transform"
monai/transforms/spatial/array.py:570: error: Signature of "__call__" incompatible with supertype "Transform"
monai/transforms/spatial/array.py:674: error: Signature of "__call__" incompatible with supertype "Transform"
monai/transforms/spatial/array.py:890: error: Signature of "__call__" incompatible with supertype "Transform"
monai/transforms/spatial/array.py:980: error: Signature of "__call__" incompatible with supertype "Transform"
monai/transforms/spatial/array.py:1064: error: Signature of "__call__" incompatible with supertype "Transform"
monai/transforms/spatial/array.py:1161: error: Signature of "__call__" incompatible with supertype "Transform"
monai/transforms/post/array.py:41: error: Signature of "__call__" incompatible with supertype "Transform"
monai/transforms/post/array.py:72: error: Signature of "__call__" incompatible with supertype "Transform"
monai/transforms/post/array.py:121: error: Signature of "__call__" incompatible with supertype "Transform"
monai/transforms/croppad/array.py:56: error: Signature of "__call__" incompatible with supertype "Transform"
```
2. `MYPYPATH=$(pwd)/monai mypy $(pwd)`

**Expected behavior**
Clean static testing.

### PR DESCRIPTION
**Pull request by benjamin-gorman**
Co-authored-by: Cameron Trentz <cameron-trentz@uiowa.edu>
Co-authored-by: Hans Johnson <hans-johnson@uiowa.edu>

Fixes #495
```
error: Signature of "__call__" incompatible with supertype "Transform"
```

### Description
```python
class Transform(ABC):
    @abstractmethod
    def __call__(self, data, *args, **kwargs):
        raise NotImplementedError
```
- first parameter must be named ``data``
  - ``img`` must be renamed
  - ``data_array`` must be renamed
- ``*args`` and ``**kwargs`` must be included
  - ``validate_kwargs`` added to ensure correctness

### Status
**Ready**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [x] Docstrings/Documentation updated

### CODE REVIEW

**Comment by Nic-Ma** [comment]
Hi @benjamin-gorman ,

Thanks for your contribution.
May I know why change to this `produce_args` and `reference_args` logic?

**Comment by hjmjohnson** [comment]
@Nic-Ma The derived classes `__call__` function does not match the signature of the superclass.

**Comment by hjmjohnson** [comment]
Superceeded by #521.

**Comment by hjmjohnson** [comment]
@benjamin-gorman Will you please close this PR?

File: monai/transforms/post/array.py [hunk 1]
```diff
@@ -69,22 +87,39 @@ def __init__(self, sigmoid: bool = False, softmax: bool = False, other: Optional
         self.softmax = softmax
         self.other = other
 
-    def __call__(  # type: ignore # see issue #495
-        self, img, sigmoid: Optional[bool] = None, softmax: Optional[bool] = None, other: Optional[Callable] = None
-    ):
+    def __call__(self, data: Iterable, *args, **kwargs):
+        """
```
Comment (hjmjohnson) [inline monai/transforms/post/array.py:91]: @benjamin-gorman These comments are good, and could/should be added to the solution identified in #521   I would recommend manually copying them into the correct places and make separate documentation only PR.

**Pull request closed without merging**